### PR TITLE
[SYCL] Improve range rounding heuristics.

### PR DESCRIPTION
**Pull request by rdeodhar**
This change avoids doing range rounding on the host device and adjusts heuristics to make rounding more likely to be profitable on CPU and GPU devices.

Signed-off-by: rdeodhar <rajiv.deodhar@intel.com>

### CODE REVIEW

File: sycl/include/CL/sycl/handler.hpp [hunk 1]
```diff
@@ -769,15 +774,16 @@ class __SYCL_EXPORT handler {
         (KI::callsThisItem());
 
     // Perform range rounding if rounding-up is enabled
+    // and there are sufficient work-items to need rounding
     // and the user-specified range is not a multiple of a "good" value.
-    if (!DisableRounding && NumWorkItems[0] % GoodLocalSizeX != 0) {
+    if (!DisableRounding && (NumWorkItems[0] > MinRangeX) &&
```
Comment (vladimirlaz) [inline sycl/include/CL/sycl/handler.hpp:779]:
```suggestion
    if (!DisableRounding && (NumWorkItems[0] >= MinRangeX) &&
```
According to the description is should be more or equal.

File: sycl/include/CL/sycl/handler.hpp [hunk 2]
```diff
@@ -731,14 +731,19 @@ class __SYCL_EXPORT handler {
...
 #if !defined(__SYCL_EXPLICIT_SIMD__) &&                                        \
     !defined(SYCL_DISABLE_PARALLEL_FOR_RANGE_ROUNDING) &&                      \
-    SYCL_LANGUAGE_VERSION >= 202001
-    // The work group size preferred by this device.
-    // A reasonable choice for rounding up the range is 32.
-    constexpr size_t GoodLocalSizeX = 32;
+    !defined(DPCPP_HOST_DEVICE_OPENMP) &&                                      \
+    !defined(DPCPP_HOST_DEVICE_PERF_NATIVE) && SYCL_LANGUAGE_VERSION >= 202001
+    // Range should be a multiple of this for reasonable performance.
+    constexpr size_t MinFactorX = 16;
```
Comment (vladimirlaz) [inline sycl/include/CL/sycl/handler.hpp:742]: Can we have these values also configurable through environment variables?
The current values can be used as default.

**Review by vladimirlaz — CHANGES_REQUESTED** [review]

File: sycl/include/CL/sycl/handler.hpp [hunk 3]
```diff
@@ -739,11 +739,29 @@ class __SYCL_EXPORT handler {
...
     // Range should be a multiple of this for improved performance.
-    constexpr size_t GoodFactorX = 32;
+    size_t GoodFactorX = 32;
     // Range should be at least this to make rounding worthwhile.
-    constexpr size_t MinRangeX = 1024;
+    size_t MinRangeX = 1024;
+
+    // Parse optional parameters of this form:
+    // MinRound:PreferredRound:MinRange
+    char *RoundParams = getenv("SYCL_PARALLEL_FOR_RANGE_ROUNDING_PARAMS");
```
Comment (vladimirlaz) [inline sycl/include/CL/sycl/handler.hpp:750]: Can we use SYCLConfig interface here (e.g. SYCLConfig<SYCL_DEVICE_FILTER>)?
In any case it is resonable to add environment variable to sycl/doc/EnvironmentVariables.md

**Review by vladimirlaz — CHANGES_REQUESTED** [review]

File: sycl/include/CL/sycl/handler.hpp (hunk 3 above)
Comment (rdeodhar) [reply sycl/include/CL/sycl/handler.hpp:750]: This feature is supposed to improve out-of-the-box experience with SYCL parallel_for. It is not meant to be exposed to users. So I have added env var control only for internal experimentation. It should not be documented.

**Review by vladimirlaz — APPROVED** [review]

**Comment by bader** [comment]
@rdeodhar, please, fix CI checks.

**Pull request merged**